File: mobile/android/app/src/main/java/com/mappls/dic/iimm/ApiClient.kt
package com.mappls.dic.iimm

import android.content.Context
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder

class ApiException(val status: Int, message: String) : IOException(message)

class SessionStore(context: Context) {
    private val prefs = EncryptedSharedPreferences.create(
        context,
        "iimm-secure-session",
        MasterKey.Builder(context).setKeyScheme(MasterKey.KeyScheme.AES256_GCM).build(),
        EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
        EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
    )
    fun token(): String? = prefs.getString("token", null)
    fun user(): User? = prefs.getString("user", null)?.let { User.from(JSONObject(it)) }
    fun projectId(): String? = prefs.getString("last-project-id", null)
    fun saveProjectId(projectId: String?) {
        if (!projectId.isNullOrBlank()) prefs.edit().putString("last-project-id", projectId).apply()
    }
    fun save(token: String, user: JSONObject) = prefs.edit().putString("token", token).putString("user", user.toString()).apply()
    fun clear() = prefs.edit().clear().apply()
}

class ApiClient(private val sessions: SessionStore) {
    private val base = BuildConfig.API_BASE_URL.trimEnd('/')

    suspend fun array(path: String): JSONArray {
        val result = request("GET", path) as JSONArray
        if (path == "/api/projects" && result.length() > 0) {
            sessions.saveProjectId(result.optJSONObject(0)?.optString("id"))
        }
        return result
    }
    suspend fun obj(path: String): JSONObject = request("GET", path) as JSONObject
    suspend fun post(path: String, body: JSONObject = JSONObject()): Any = request("POST", path, body)
    suspend fun patch(path: String, body: JSONObject): Any = request("PATCH", path, body)

    suspend fun download(path: String): ByteArray = withContext(Dispatchers.IO) {
        val connection = URL("$base$path").openConnection() as HttpURLConnection
        try {
            connection.requestMethod = "GET"
            connection.connectTimeout = 15_000
            connection.readTimeout = 30_000
            sessions.token()?.let { connection.setRequestProperty("Authorization", "Bearer $it") }
            val status = connection.responseCode
            if (status !in 200..299) {
                val text = connection.errorStream?.bufferedReader()?.use { it.readText() }.orEmpty()
                throw ApiException(status, runCatching { JSONObject(text).optString("error") }.getOrNull().orEmpty().ifBlank { "Download failed ($status)" })
            }
            connection.inputStream.use { it.readBytes() }
        } finally { connection.disconnect() }
    }

    suspend fun login(userId: String): Session {
        val result = post("/api/auth/login", JSONObject().put("userId", userId)) as JSONObject
        val userJson = result.getJSONObject("user")
        sessions.save(result.getString("token"), userJson)
        return Session(result.getString("token"), User.from(userJson), result.optJSONObject("tenant")?.optString("name"))
    }

    suspend fun reverseGeocode(lat: Double, lng: Double): String {
        val path = "/api/mappls/reverse-geocode?lat=${enc(lat.toString())}&lng=${enc(lng.toString())}"
        return obj(path).getString("address")
    }

    suspend fun uploadMedia(
        bytes: ByteArray,
        mimeType: String,
        fileName: String,
        lat: Double,
        lng: Double,
        accuracyMeters: Double?,
    ): JSONObject = withContext(Dispatchers.IO) {
        val connection = URL("$base/api/media").openConnection() as HttpURLConnection
        try {
            connection.requestMethod = "POST"
            connection.connectTimeout = 15_000
            connection.readTimeout = 40_000
            connection.doOutput = true
            connection.setRequestProperty("Accept", "application/json")
            connection.setRequestProperty("Content-Type", mimeType)
            connection.setRequestProperty("X-File-Name", fileName)
            connection.setRequestProperty("X-Capture-Lat", lat.toString())
            connection.setRequestProperty("X-Capture-Lng", lng.toString())
            accuracyMeters?.let { connection.setRequestProperty("X-Capture-Accuracy", it.toString()) }
            connection.setRequestProperty("X-Captured-At", java.time.Instant.now().toString())
            sessions.token()?.let { connection.setRequestProperty("Authorization", "Bearer $it") }
            connection.outputStream.use { it.write(bytes) }
            val status = connection.responseCode
            val text = (if (status in 200..299) connection.inputStream else connection.errorStream)
                ?.bufferedReader()?.use { it.readText() }.orEmpty()
            if (status !in 200..299) throw ApiException(
                status,
                runCatching { JSONObject(text).optString("error") }.getOrNull().orEmpty()
                    .ifBlank { "Evidence upload failed ($status)" },
            )
            JSONObject(text)
        } finally { connection.disconnect() }
    }

    private suspend fun request(method: String, path: String, body: JSONObject? = null): Any = withContext(Dispatchers.IO) {
        val connection = URL("$base$path").openConnection() as HttpURLConnection
        try {
            connection.requestMethod = method
            connection.connectTimeout = 15_000
            connection.readTimeout = 25_000
            connection.setRequestProperty("Accept", "application/json")
            sessions.token()?.let { connection.setRequestProperty("Authorization", "Bearer $it") }
            if (body != null) {
                connection.doOutput = true
                connection.setRequestProperty("Content-Type", "application/json")
                connection.outputStream.use { it.write(body.toString().toByteArray()) }
            }
            val status = connection.responseCode
            val text = (if (status in 200..299) connection.inputStream else connection.errorStream)?.bufferedReader()?.use { it.readText() }.orEmpty()
            if (status !in 200..299) throw ApiException(status, runCatching { JSONObject(text).optString("error") }.getOrNull().orEmpty().ifBlank { "Request failed ($status)" })
            val trimmed = text.trim()
            if (trimmed.startsWith("[")) JSONArray(trimmed) else JSONObject(trimmed.ifBlank { "{}" })
        } finally { connection.disconnect() }
    }

    private fun enc(value: String) = URLEncoder.encode(value, Charsets.UTF_8.name())
}

File: mobile/ios/IIMM/APIClient.swift
import Foundation
import Security

enum APIError: LocalizedError {
  case server(String)
  var errorDescription: String? {
    if case .server(let value) = self { return value }
    return "Request failed"
  }
}

final class KeychainSession {
  private let service = "com.mappls.dic.iimm.session", account = "bearer"
  func token() -> String? {
    let query: [String: Any] = [
      kSecClass as String: kSecClassGenericPassword, kSecAttrService as String: service,
      kSecAttrAccount as String: account, kSecReturnData as String: true,
    ]
    var item: CFTypeRef?
    guard SecItemCopyMatching(query as CFDictionary, &item) == errSecSuccess,
      let data = item as? Data
    else { return nil }
    return String(data: data, encoding: .utf8)
  }
  func save(_ token: String) {
    clear()
    let data = Data(token.utf8)
    SecItemAdd(
      [
        kSecClass as String: kSecClassGenericPassword, kSecAttrService as String: service,
        kSecAttrAccount as String: account, kSecValueData as String: data,
      ] as CFDictionary, nil)
  }
  func clear() {
    SecItemDelete(
      [
        kSecClass as String: kSecClassGenericPassword, kSecAttrService as String: service,
        kSecAttrAccount as String: account,
      ] as CFDictionary)
  }
}

final class APIClient {
  static let base = URL(string: "https://mappls-dic-iimm.replit.app")!
  let keys = KeychainSession()

  func get(_ path: String) async throws -> Any { try await request(path: path, method: "GET") }
  func post(_ path: String, body: [String: Any] = [:]) async throws -> Any {
    try await request(path: path, method: "POST", body: body)
  }
  func patch(_ path: String, body: [String: Any]) async throws -> Any {
    try await request(path: path, method: "PATCH", body: body)
  }
  func download(_ path: String) async throws -> Data {
    var request = URLRequest(url: URL(string: path, relativeTo: Self.base)!)
    request.timeoutInterval = 30
    if let token = keys.token() { request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
    let (data, response) = try await URLSession.shared.data(for: request)
    guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
      throw APIError.server("Report download failed")
    }
    return data
  }
  func login(userId: String) async throws -> Session {
    guard
      let result = try await post("/api/auth/login", body: ["userId": userId]) as? [String: Any],
      let token = result["token"] as? String, let userJSON = result["user"] as? [String: Any],
      let user = User(userJSON)
    else { throw APIError.server("Invalid login response") }
    keys.save(token)
    let tenant = (result["tenant"] as? [String: Any])?["name"] as? String
    return Session(token: token, user: user, tenantName: tenant)
  }
  func reverse(lat: Double, lng: Double) async throws -> String {
    let result =
      try await get("/api/mappls/reverse-geocode?lat=\(lat)&lng=\(lng)") as? [String: Any]
    return result?["address"] as? String ?? "\(lat), \(lng)"
  }

  func upload(
    data: Data, mimeType: String, fileName: String, lat: Double, lng: Double,
    accuracyMeters: Double?
  ) async throws -> [String: Any] {
    var request = URLRequest(url: URL(string: "/api/media", relativeTo: Self.base)!)
    request.httpMethod = "POST"
    request.timeoutInterval = 45
    request.setValue("application/json", forHTTPHeaderField: "Accept")
    request.setValue(mimeType, forHTTPHeaderField: "Content-Type")
    request.setValue(fileName, forHTTPHeaderField: "X-File-Name")
    request.setValue(String(lat), forHTTPHeaderField: "X-Capture-Lat")
    request.setValue(String(lng), forHTTPHeaderField: "X-Capture-Lng")
    request.setValue(ISO8601DateFormatter().string(from: Date()), forHTTPHeaderField: "X-Captured-At")
    if let accuracyMeters { request.setValue(String(accuracyMeters), forHTTPHeaderField: "X-Capture-Accuracy") }
    if let token = keys.token() { request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
    request.httpBody = data
    let (responseData, response) = try await URLSession.shared.data(for: request)
    guard let http = response as? HTTPURLResponse else { throw APIError.server("No server response") }
    let value = try JSONSerialization.jsonObject(with: responseData)
    guard (200..<300).contains(http.statusCode), let object = value as? [String: Any] else {
      throw APIError.server((value as? [String: Any])?["error"] as? String ?? "Evidence upload failed (\(http.statusCode))")
    }
    return object
  }

  func parseGisFile(data: Data, fileName: String) async throws -> [String: Any] {
    var request = URLRequest(url: URL(string: "/api/gis/parse-file", relativeTo: Self.base)!)
    request.httpMethod = "POST"
    request.timeoutInterval = 60
    request.setValue("application/json", forHTTPHeaderField: "Accept")
    request.setValue("application/octet-stream", forHTTPHeaderField: "Content-Type")
    request.setValue(fileName, forHTTPHeaderField: "X-File-Name")
    if let token = keys.token() { request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
    request.httpBody = data
    let (responseData, response) = try await URLSession.shared.data(for: request)
    guard let http = response as? HTTPURLResponse else { throw APIError.server("No server response") }
    let value = try JSONSerialization.jsonObject(with: responseData)
    guard (200..<300).contains(http.statusCode), let object = value as? [String: Any] else {
      throw APIError.server((value as? [String: Any])?["error"] as? String ?? "GIS file parsing failed (\(http.statusCode))")
    }
    return object
  }

  private func request(path: String, method: String, body: [String: Any]? = nil) async throws -> Any
  {
    var request = URLRequest(url: URL(string: path, relativeTo: Self.base)!)
    request.httpMethod = method
    request.timeoutInterval = 25
    request.setValue("application/json", forHTTPHeaderField: "Accept")
    if let token = keys.token() {
      request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
    }
    if let body {
      request.setValue("application/json", forHTTPHeaderField: "Content-Type")
      request.httpBody = try JSONSerialization.data(withJSONObject: body)
    }
    let (data, response) = try await URLSession.shared.data(for: request)
    guard let http = response as? HTTPURLResponse else {
      throw APIError.server("No server response")
    }
    let value = try JSONSerialization.jsonObject(with: data)
    guard (200..<300).contains(http.statusCode) else {
      throw APIError.server(
        (value as? [String: Any])?["error"] as? String ?? "Request failed (\(http.statusCode))")
    }
    return value
  }
}
